### PR TITLE
#41196  | Add aws KMS encryption in the elastic search repository-s3 plugin

### PR DESCRIPTION
**Pull request by Prasannasaraf**

### CODE REVIEW

File: plugins/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3BlobStore.java [hunk 1]
```diff
@@ -49,16 +53,26 @@
...
         this.bufferSize = bufferSize;
         this.cannedACL = initCannedACL(cannedACL);
         this.storageClass = initStorageClass(storageClass);
         this.repositoryMetaData = repositoryMetaData;
+        this.sseAwsKeyId = sseAwsKeyId;
+    }
+
+    public boolean sseAwsKeyIsEmpty() {
+        if (sseAwsKeyId == null || sseAwsKeyId.trim().isEmpty()) {
+            return true;
```
Comment (satishdash) [inline plugins/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3BlobStore.java]: Can be simplified into:
```
return sseAwsKeyId == null || sseAwsKeyId.trim().isEmpty()
```

**Review by satishdash — CHANGES_REQUESTED** [review]

**Review by satishdash — APPROVED** [review]

**Review by original-brownbear — CHANGES_REQUESTED** [review]
Thanks for this PR @Prasannasaraf !

I left a few comments/questions on the exact mechanics of how the new setting should work.

Also, we definitely have to add a 3rd party test for this that runs if the appropriate environment variable is set. Something along the lines of what we did for Azure SAS tokens is probably what we want here (see https://github.com/elastic/elasticsearch/pull/42982/files#diff-972bd167c62a9bf3a9bf9a80d6ea2298R79).

I also wonder if we sohuld make setting an `endpoint` on a repo or client mutually exclusive with using KMS on the settings layer already since this won't work with non-AWS S3 endpoints anyway?

File: plugins/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3Repository.java [hunk 2]
```diff
@@ -75,6 +75,12 @@
      */
     static final Setting<Boolean> SERVER_SIDE_ENCRYPTION_SETTING = Setting.boolSetting("server_side_encryption", false);
 
+    /**
+     * When aws_kms_key is set, files are encrypted on server side using kms key provided or AE256 algorithm is used.
+     * Defaults to AES256 algorithm.
+     */
+    static final Setting<String> AWS_KMS_KEY = Setting.simpleString("aws_kms_key");
```
[the remark this answers is not in the document]
Comment (tlrx) [reply plugins/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3Repository.java]: If we want to add support for KMS then I think we need smarter settings that allow to distinguish between server side encryption using AES256, SSE with KMS and default key and SSE with KMS and specific key id.

> I'd also like a second opinion here (@ywelsch, @tlrx maybe?) on whether or not this should be a secure setting.

I tend to think that it should be a secured setting. I also think that it should be correctly documented as we don't want users to remove or update keys and not being able to decrypt data afterwards? Also, @albertzaharovits might have an opinion on this.

File: plugins/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3BlobContainer.java [hunk 3]
```diff
@@ -313,14 +314,22 @@ void executeSingleUpload(final S3BlobStore blobStore,
         }
 
         final ObjectMetadata md = new ObjectMetadata();
+        final PutObjectRequest putRequest = new PutObjectRequest(blobStore.bucket(), blobName, input, md);
         md.setContentLength(blobSize);
         if (blobStore.serverSideEncryption()) {
```
[the remark this answers is not in the document]
Comment (Prasannasaraf) [reply plugins/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3BlobContainer.java]: Should we put a warning if sse is disabled but KMS key is set and continue?

File: plugins/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3Repository.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (albertzaharovits) [reply plugins/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3Repository.java]: I don't think the key identifier should be a secure setting. It's just an identifier, and I think that putting it in the keystore creates unnecessary complications for the user.

[the remark this answers is not in the document]
Comment (albertzaharovits) [reply plugins/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3Repository.java]: Also, from https://docs.aws.amazon.com/AmazonS3/latest/API/API_PutObject.html I see that the API has complexish rules:
```
x-amz-server-side​-encryption
x-amz-server-side-encryption-aws-kms-key-id
x-amz-server-side-encryption-context

Note:
If you specify x-amz-server-side-encryption:aws:kms, but don't provide x-amz-server-side- encryption-aws-kms-key-id, Amazon S3 uses the default AWS KMS key to protect the data.
```

that don't really make much sense to me.

I would think to whitelist these request headers, together with `x-amz-server-side​-encryption​-customer-algorithm`, `x-amz-server-side​-encryption​-customer-key`, `x-amz-server-side​-encryption​-customer-key-MD5` from below, and delegate the responsibility to the user to validate them. What do you think?

**Comment by original-brownbear** [comment]
Hi @Prasannasaraf,

sorry for the delay here. As I said in my last comment, this will require adding some actual third party tests against real AWS S3 to validate that it works. Are you interested in working on those and continuing here? (I realize that this is a non-trivial amount of effort on top of the actual change so no worries if not.)

**Comment by Prasannasaraf** [comment]
Hi @original-brownbear 

I can work on adding third-party tests along with the changes suggested by @albertzaharovits. Do let me know the amount of work involved here.

**Comment by jstolle** [comment]
Is there an ETA on the completion of this merge?

**Comment by wywywywy** [comment]
Is this ever going to get merged?

**Pull request closed without merging**